salt/states/grains.py

```diff
@@ -44,6 +44,13 @@ def exists(name, delimiter=DEFAULT_TARGET_DELIM):
         ret['comment'] = 'Grain does not exist'
     return ret
 
+def flatten(li, flattened = list()):
+    for subli in li:
+        if type(subli) == list:
+            flatten(subli, flattened)
+        else:
+            flattened.append(frozenset(subli))
+    return set(flattened)
```

Comment:
Would it be better to just use `set` here? The default arg *definitely* should be changed, though.

I'm thinking something like:

```
def flatten(li, flattened=None):
    flattened = flattened or set()
    for subli in li:
        if type(subli) == list:
            flattened.update(flatten(subli))
        else:
            flattened.add(frozenset(subli))
    return flattened
```

I'm not entirely certain that's the best approach either, though.

Reply:
Great, I will fix those things :)